drivers/mpu9250/MPU9250.cpp

```diff
@@ -250,8 +312,26 @@ int MPU9250::mpu9250_deinit()
 	return 0;
 }
 
+int MPU9250::pinThread(int cpu)
```

Comment:
I think that's answering a different concern from mine.

The kernel's spi thread is in a different scheduling context, so the way it competes with runnable processes is up to the kernel to decide. Assigning IRQ affinity, and moving it to the SCHED_FIFO scheduler will definitely give it priority over most user processes, and the kernel is also free to deal with other runnable processes in the usual (for Linux) ways.

I'm not sure there's a comparable implementation in use for NuttX, tho. From what I've been hearing (I'm still grokking the code), workqueues in NuttX sound a lot like Linux's single-thread workqueue, where workers are run sequentially and with strict serialization.

If that's true, then if a NuttX "process" creates its own thread, that thread will be scheduled with rules applied to it vs. ALL the workers on the workqueue, who by definition have the same priority because of the aforementioned serialization. So once your thread gets the CPU, all other work will stop until you're done, even if a more "urgent" worker is already queued.

That's great for you, but might not be great for others.

My sense is that the singlegthread workqueue in NuttX is the root problem. Having a threaded device driver works around the problem for that driver, but doesn't fix the overall situation because the scheduler still doesn't have enough insight to prioritize workers correctly in a global sense.

I think that's the point @julianoes was making. Or at least, it's mine.